checker/tests/run-pass/contract_annotations.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+//
+
+// Tests for annotations from the contracts crate.
+
+// TODO: those tests currently do not do anything because all relevant functions are marked
+// as angelic. Once this is not longer the case, real assertion logic should be added.
+
+use contracts::*;
+use mirai_annotations::*;
+
+pub fn main() {
+    use_pre_post();
+    use_trait();
+    use_invariant();
+}
+
+// Simple pre/post
+// ---------------
+
+#[pre(x > 0)]
+#[post(ret >= x)]
+pub fn pre_post(x: i32) -> i32 {
+    return x;
+}
+
+fn use_pre_post() {
+    verify!(pre_post(1) >= 1);
+}
+
+// Trait pre/post
+
+#[contract_trait]
+trait Adder {
+    fn get(&self) -> i32;
+
+    #[pre(x > 0)]
+    #[pre(self.get() <= std::i32::MAX - x)]
+    #[post(ret == old(self.get()) && self.get() > old(self.get()))]
+    fn get_and_add(&mut self, x: i32) -> i32;
+}
+
+struct MyAdder {
+    x: i32,
+}
+
+#[contract_trait]
+impl Adder for MyAdder {
+    fn get(&self) -> i32 {
+        self.x
+    }
+    fn get_and_add(&mut self, x: i32) -> i32 {
+        let c = self.x;
+        // The below is currently needed because in MIRAI compilation mode, the runtime
+        // assertion generated by the matching #[pre] is dead code. Therefore the compiler
+        // does not know that this overflow is protected and issues a warning.
+        // TODO: find a way to fix this as it is a general problem
+        assert!(self.get() <= std::i32::MAX - x);
```

Comment:
assert! is not part of the MIRAI contract language. You probably want to use checked_assume! here.

Reply:
Please look at the comment. The problem seems to be that the checked_assume (which the contracts already generate -- see pre conditions) is compiled into:

  if cfg!("mirai") { assume! ... } else { assert! ... }

When MIRAI is on the assert! will be dead code and rustc doesn't know that overflow has been checked for.

At least that is my explanation so far for this behavior.

Reply:
When mirai has been specified then MIRAI == rustc and the assume! should be sufficient. Of course, my assumption has always been that when mirai is specified, the produced code is not actually run.

Reply:
Also, checked_assume! is equivalent to assert! when mirai is not specified.

Reply:
I understand what checked_assume et. al do. I was assuming that actually the rustc checker generates this warning, because it flags the according method as angelic, but now I understand what is happening. Simplified example:

  #[pre(std::i32::MAX - x >= y)]
  fn(x,y) { x + y }

  ====v=====

  fn(x,y) {
    checked_precondition(std::i32::MAX - x >= y)]
    let f = || x + y;
    let ret = f();
    ...
    return ret;
} 

We don't seem to propagate any knowledge about constraints on x and y into the closure fn.

I changed the TODO in the code to document this problem.

Reply:
The way things are supposed to work is that the closures get inferred pre-conditions and the invocation of the closure satisfies them with the specified precondition.

Reply:
Well that seems not to work. I added a minimal repro in a new test case `closure_precond_inference.rs` as part of this PR.

I think this PR here is good to go but waiting for your approval.

Reply:
The PR has to pass the CI first.

Reply:
Passing now. I wish travis would send me an email...